megalista_dataflow/data_sources/file/file_data_source.py

```diff
@@ -0,0 +1,185 @@
+# limitations under the License.
+
+"""
+This module is responsible for handling (read, write) data from files. As of now, it handles CSV and Parquet files as data sources.
+It depends on file_provider module, which is responsible for handling file operations.
+"""
+
+import pandas as pd
+import io
+import os
+from datetime import datetime, timedelta
+from typing import Any, List, Iterable, Tuple, Dict
+
+import logging
+
+from models.execution import DestinationType, Execution, Batch, TransactionalType
+from models.options import DataflowOptions
+
+
+from data_sources.base_data_source import BaseDataSource
+from data_sources.file.file_provider import FileProvider
+import data_sources.data_schemas as DataSchemas
+
+_LOGGER_NAME = 'megalista.data_sources.File'
+
+
+class FileDataSource(BaseDataSource):
+    def __init__(self, transactional_type: TransactionalType, dataflow_options: DataflowOptions, destination_type: DestinationType):
+        self._transactional_type = transactional_type
+        self._dataflow_options = dataflow_options
+        self._destination_type = destination_type
+  
+    def retrieve_data(self, execution: Execution) -> Iterable[Tuple[Execution, Dict[str, Any]]]:
+        if self._transactional_type == TransactionalType.NOT_TRANSACTIONAL:
+            return self._retrieve_data_non_transactional(execution)
+        else:
+            return self._retrieve_data_transactional(execution)
+  
+    def _retrieve_data_non_transactional(self, execution: Execution) -> Iterable[Tuple[Execution, Dict[str, Any]]]:
+        # Get Data Source
+        data_source = self._get_data_source(execution.source.source_metadata[0])
+        # Get Data Frame
+        df = data_source.get_data_frame(execution.source.source_metadata[1])
+        if df is not None:
+            # Process Data Frame
+            for _, row in df.iterrows():
+                yield execution, FileDataSource._convert_row_to_dict(row)
+        else:
+            raise Exception("Unable to read from data source.")
+  
+    def _retrieve_data_transactional(self, execution: Execution) -> Iterable[Tuple[Execution, Dict[str, Any]]]:
+        # Get Data Source
+        data_source = self._get_data_source(execution.source.source_metadata[0])
+        # Get Data Frame
+        df = data_source.get_data_frame(execution.source.source_metadata[1])
+        # Get Uploaded Data Frame
+        df_uploaded = data_source.get_data_frame(execution.source.source_metadata[1], is_uploaded=True)
+        
+        if df is not None:
+            # Get items that haven't been processed yet
+            df_merged = df.merge(df_uploaded, how='outer')
+            df_distinct = df_merged.drop(df_merged[df_merged.timestamp.notnull()].index)
+            # Process Data Frame
+            for index, row in df_distinct.iterrows():
+                yield execution, FileDataSource._convert_row_to_dict(row)
+        else:
+            raise Exception("Unable to read from data source.")
```

Comment:
I think we should include the Source Name here (key) to indicate what source wasn't found.